### PR TITLE
Implement selectable and editable text

### PR DESCRIPTION
**Pull request by Dinnerbone**
The initial approach will be limited to ASCII support as we have lots of issues with encoding right now. Also don't expect many keyboard shortcuts or quality of life - basic support unlocks lots of stuff, so I'm focusing on bare minimum right now.

Todo:
- [x] Implement all of Selection (avm1) - #271
- [x] Allow typing of text and inserting it into the edit box
- [x] Show a caret or highlight selected text
- [x] Focus tracking
- [x] Focus on clicking text fields
- [x] Focus on script access
- [x] Allow arrow keys
- [x] Allow backspace & delete shortcuts
- [x] Limit editing to only editable texts
- [x] Testing ??

### CODE REVIEW

File: core/src/display_object/edit_text.rs [hunk 1]
```diff
@@ -831,6 +931,118 @@ impl<'gc> EditText<'gc> {
                 .firing_variable_binding = false;
         }
     }
+
+    pub fn get_selection(self) -> Option<TextSelection> {
+        self.0.read().selection
+    }
```
Comment (Herschel) [inline core/src/display_object/edit_text.rs:937]: Nit: Rust naming convention should be `selection` here (I know we're not very consistent with this yet).

**Comment by Toad06** [comment]
Thanks for this PR, very cool.
Many things work like a charm now. 👍

Here are some feedbacks, the last two ones are possibly unrelated to this PR:

- Backspace and Delete shortcuts don't work on the Web backend.

- Focus remains on the text field even after you've clicked on another element.

- In [Soap Bubble 1](https://www.newgrounds.com/portal/view/52537/), when you enter a valid password (eg. "ldkd"), the game should get you to the expected level but nothing happens here (could this be because of the newline thing you've talked about on Discord?).

- In [Super Chick Sisters](https://www.newgrounds.com/portal/view/396372), what you type won't be displayed on screen but the password ("GOPAM") will work.

**Comment by Herschel** [comment]
That reminds me: in Flash, if a font is using outline glyphs and doesn't have certain glyphs embedded, those characters can't be entered in the text field. For example, I wouldn't be able to type "e" if the "e" character was not embedded in the font. Currently this PR allows these characters to be entered; they don't appear but still exist in the text.

File: core/src/avm1/globals/text_field.rs [hunk 2]
```diff
@@ -528,6 +564,23 @@ pub fn attach_virtual_properties<'gc>(
...
+            Some(fn_proto),
+            fn_proto,
+        ),
+        Some(FunctionObject::function(
+            gc_context,
+            Executable::Native(set_selectable),
+            Some(fn_proto),
+            fn_proto,
+        )),
+        ReadOnly.into(),
```
[the remark this answers is not in the document]
Comment (Dinnerbone) [reply core/src/avm1/globals/text_field.rs:582]: It seems to be deletable for me. It shows up in an enum until I delete it (from the proto) and then it doesn't.

**Comment by Dinnerbone** [comment]
>     * Backspace and Delete shortcuts don't work on the Web backend.
Thanks, fixed! Not very proud of the fix, but it'll work for now.

> 
>     * Focus remains on the text field even after you've clicked on another element.
Hmmm, got an example of this?

> 
>     * The caret always seems to be invisible when there's no text in the field; and always black otherwise. This can be seen in [Bouncy The Ball](https://www.newgrounds.com/portal/view/204189) (password section).
Fixed the invisible caret

>     * In [Soap Bubble 1](https://www.newgrounds.com/portal/view/52537), when you enter a valid password (eg. "ldkd"), the game should get you to the expected level but nothing happens here (could this be because of the newline thing you've talked about on Discord?).
Not sure why, it doesn't seem to have any extra characters. It may not be because of this, but I'll try and debug.

>     * In [Super Chick Sisters](https://www.newgrounds.com/portal/view/396372), what you type won't be displayed on screen but the password ("GOPAM") will work.
That's the newline bug - if you hit backspace you'll delete that newline and see your text. It looks like they're matching after stripping the newline, so it worked even though it's not exactly correct.

**Comment by Toad06** [comment]
👍

> Hmmm, got an example of this?

In Soap Bubble for instance: click "Enter code", click the text input and type something. Click anywhere else on the file, the input will still have the focus (and display what you type on the keyboard).


For the passwords not being recognised, it seems that _Bouncy The Ball_ has a similar issue (eg. "bounce" should be a valid password but the game doesn't accept it).

**Comment by Dinnerbone** [comment]
Hmm, that's odd. I can't make a movie to reproduce that behaviour, when I click out of a text box it always seems to keep the focus on the text box

🤔

**Comment by Toad06** [comment]
Oh, wow! I've checked with Flash on Chrome, IE and the projector and none of these versions kept the focus on the text box in such situation. Flash's insane sometimes. 😅

**Comment by Dinnerbone** [comment]
Sorry maybe I wasn't clear - I do see that behaviour (losing focus) in that game, but I can't make my own test case reproduce it. Whatever I try makes it stick the focus. That game is doing something different but I'm not sure what?

**Comment by Dinnerbone** [comment]
Looks like Soap Bubble binds the text to `_root.levelcode` and that seems to be broken with our current variable binding - we read it but can't seem to write to it.

**Comment by Dinnerbone** [comment]
@Toad06 with the exception of the focus thing, all those issues should be fixed now!

**Comment by Toad06** [comment]
I've checked a few other files using password systems, everything's perfect. 👍

**Comment by Herschel** [comment]
Thanks!

**Comment by Poopooracoocoo** [comment]
fyi caret colour is always black :)

**Comment by z3ntu** [comment]
Can confirm this works great for the game Sim Taxi, thanks!
I've only noticed that the cursor is on the left side of the input field when no text is entered but then when entering text it's centered; so basically the cursor should probably start in the center when the text is supposed to be centered.

**Pull request merged**